templates/CMakeLists.txt.template

```diff
@@ -416,6 +452,67 @@
     endforeach()
   endfunction()
 
+  #  protobuf_generate_grpc_cpp_with_import_path
+  #  --------------------------
+  #
+  #   Add custom commands to process ``.proto`` files to C++ using protoc and
+  #   GRPC plugin::
+  #
+  #     protobuf_generate_grpc_cpp_with_import_path <FILE_LOCATION> <IMPORT_PATH>
+  #
+  #   ``FILE_LOCATION``
+  #     The relative path of the ``.proto`` file to the project root
+  #   ``IMPORT_PATH``
+  #     The proto import path that itself expected to be placed in. For
+  #     example, a "bar.proto" file wants to be imported as
+  #     `import "foo/bar.proto"`. Then we should place it under
+  #     "<ProtoBuf_Include_Path>/foo/bar.proto" instead of
+  #     "<ProtoBuf_Include_Path>/third_party/foo/bar.proto". This ensures
+  #     correct symbol being generated and C++ include path being correct.
+  #     More info can be found at https://github.com/grpc/grpc/pull/25272.
+  #
+  function(protobuf_generate_grpc_cpp_with_import_path FILE_LOCATION IMPORT_PATH)
```

Comment:
Looks like this functions is 95% the same as the original `protobuf_generate_grpc_cpp` you copied.
Instead of duplicating the code (bad for maintainability), can you adjust the original function to be able to handle the case that requires putting protos into the centralized location?

Btw, I think many of our .proto in the grpc/grpc repo had their import statements modified to workaround the same issue as you're trying to workaround here.
Perhaps there should be some cleanup afterwards to make sure the import statements in all .proto files (inside and outside of grpc/grpc) are used consistently?

Reply:
Also, in the current form, it's quite non-obvious what's actually different about the `protobuf_generate_grpc_cpp` vs `protobuf_generate_grpc_cpp_with_import_path`. It would be better to use a single function everywhere and have only one copy for "protobuf_generate" function (but set different args depending on what our needs are).

Reply:
To Illustrate the hacks we've done to the .proto files in our repository, see the same proto in grpc/grpc and grpc-java:

https://github.com/grpc/grpc/blob/61b34dfaee80d36b505fbb488cc73b2afd803c2e/src/proto/grpc/testing/control.proto#L17

https://github.com/grpc/grpc-java/blob/78ccc81fd5c3a9c94c8a202f141e477b946a263b/benchmarks/src/main/proto/grpc/testing/control.proto#L16

as you can see, grpc-java uses the "canonical import path" (`grpc/testing/payloads.proto` "as in google3"), but C-core uses the tweaked one: `src/proto/grpc/testing/payloads.proto`.

Reply:
New method renamed to `protobuf_generate_grpc_cpp_with_import_path_correction`, the old method will be a thin wrapper around the new method as suggested. Manually tested that the old method works (able to compile existing protos under src/proto).

For user's usage (https://github.com/search?q=protobuf_generate_grpc_cpp&type=code):

* gRPC as Submodule: https://github.com/IvanSafonov/grpc-cmake-example
* ProtoBuf provides a more generic command: https://www.f-ax.de/dev/2020/11/08/grpc-plugin-cmake-support.html
* Copied the implementation ([example](https://github.com/sandvikcode/pubsuber/blob/1083c20e71fc057accf1bc0e7fbb31008efaf885/cmake/grpc.cmake))

As discussed, we might want to continue our support for the submodule use case, so keeping the original signature might be worthwhile. I didn't find any usage of using this method from the CMake gRPC target (probably because we don't support this approach yet).

Given that the ProtoBuf repo did provide a plugin command, we can recommend users to use that instead of our CMake "internal" command.

---

For the import path tweak, I will work on that in follow-up PRs.